File: sdk/src/test/java/com/cxense/cxensesdk/UserProviderTest.java
package com.cxense.cxensesdk;

import com.cxense.cxensesdk.model.ContentUser;

import org.junit.Before;
import org.junit.Test;
import org.powermock.reflect.Whitebox;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.Mockito.verify;
import static org.powermock.api.mockito.PowerMockito.doReturn;
import static org.powermock.api.mockito.PowerMockito.mock;
import static org.powermock.api.mockito.PowerMockito.spy;

/**
 * @author Dmitriy Konopelkin (dmitry.konopelkin@cxense.com) on (2018-10-15).
 */
public class UserProviderTest extends BaseTest {
    private AdvertisingIdProvider advertisingIdProvider;
    private UserProvider userProvider;

    @Before
    public void setUp() throws Exception {
        super.setUp();
        advertisingIdProvider = mock(AdvertisingIdProvider.class);
        userProvider = spy(new UserProvider(advertisingIdProvider));
    }

    @Test
    public void getUserId() {
        userProvider.getUserId();
        verify(advertisingIdProvider).getDefaultUserId();
    }

    @Test
    public void getUserIdFilledBefore() {
        String id = "VeryVeryVeryGoodId";
        Whitebox.setInternalState(userProvider, "userId", id);
        assertEquals(id, userProvider.getUserId());
    }

    @Test
    public void setUserId() {
        String id = "VeryVeryVeryGoodId";
        userProvider.setUserId(id);
        assertEquals(id, Whitebox.getInternalState(userProvider, "userId"));
    }

    @Test(expected = IllegalArgumentException.class)
    public void setUserIdNull() throws Exception {
        userProvider.setUserId(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void setUserIdBad() throws Exception {
        userProvider.setUserId("BadId");
    }

    @Test
    public void getContentUser() {
        String id = "VeryVeryVeryGoodId";
        doReturn(id).when(userProvider).getUserId();
        assertNotNull(userProvider.getContentUser());
        verify(userProvider).getUserId();
    }

    @Test
    public void getContentUserFilledBefore() {
        ContentUser user = mock(ContentUser.class);
        Whitebox.setInternalState(userProvider, "defaultUser", user);
        assertEquals(user, userProvider.getContentUser());
    }
}

File: sdk/src/test/java/io/piano/android/cxense/UserProviderTest.kt
package io.piano.android.cxense

import org.mockito.kotlin.mock
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import java.util.UUID
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull

class UserProviderTest {
    private val advertisingIdProvider: AdvertisingIdProvider = mock()
    private val provider = UserProvider(advertisingIdProvider)

    @Test
    fun getUserId() {
        whenever(advertisingIdProvider.defaultUserId).thenReturn(USER_ID)
        assertEquals(USER_ID, provider.userId)
        assertEquals(USER_ID, provider.userId)
        verify(advertisingIdProvider, times(1)).defaultUserId
    }

    @Test
    fun getUserIdNotAaid() {
        whenever(advertisingIdProvider.defaultUserId).thenReturn(null)
        assertNotNull(UUID.fromString(provider.userId))
    }

    @Test
    fun setInvalidUserId() {
        assertFailsWithMessage<IllegalArgumentException>("user id must be", "Expected fail for user id") {
            provider.userId = "1"
        }
    }

    companion object {
        private const val USER_ID = "test user id"
    }
}
